# Part 1. Equation of a Slime

How many late days are you using for this assignment?

1 late day

In [9]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, accuracy_score
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [10]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

# Using pandas to load the dataset
data = pd.read_csv('science_data_large.csv')  # Ensure the CSV file path is correct

# Output the first 15 rows of the data
print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(data.info())
print(data.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [11]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

# Splitting the dataset into features (X) and label (y)
X = data[['Mols KCL', 'Temperature °C']]  # Adjust column names as needed
y = data['Size nm^3']  # Adjust based on how the target is defined

# Use sklearn to split the features and labels into a training/test set (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [ ]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = np.array([[403, 694]])  # Example: KCl: 500, Temperature: 700
predicted_size = model.predict(sample_data)
print(f'The predicted size for KCl: 500 and Temperature: 700 is: {predicted_size[0]}')

# Report the score for that model (R^2 score)
score = model.score(X_test, y_test)
print(f'Model score (R^2): {score}')

# Extract the coefficients and intercept from the model
coefficients = model.coef_
intercept = model.intercept_
print(f'Coefficients: {coefficients}, Intercept: {intercept}')


The predicted size for KCl: 500 and Temperature: 700 is: 607890.2430904816
Model score (R^2): 0.8552472077276096
Coefficients: [1032.69506649  866.14641337], Intercept: -409391.4795834087


C:\Users\<redacted>\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

$Size = -409391.48 + (1032.7 * Temperature) + (866.15 * KCl)$

Report on score and explain meaning:

The R^2 score is a measure of the correlation between what the model predicts and the actual results. An R^2 score of 1 means the model always perfectly predicts the output, while an R^2 score of 0 means the model is not well fit. Since the score was 0.8552472077276096, it means the model is fairly accurate.

## 4. Use Cross Validation

In [13]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance

# Create the KFold object with shuffling enabled
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation using the KFold instance 
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='r2')
print(f'Cross-validation scores: {cv_scores}')
print(f'Average CV score: {np.mean(cv_scores)}, Standard deviation: {np.std(cv_scores)}')

Cross-validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Average CV score: 0.8597294202684646, Standard deviation: 0.018387737139306443


Write findings here:

The scores from the cross validation are all somewhat similar, with a standard deviation of only ~0.0184. Since the mean value is close to 1, and the standard deviation is close to 0, this means that the model should be fairly consistient at making predictions across a viriety of data sets.

## 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.

# Using the PolynomialFeatures library to create an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)  # Transforming the feature set

# Initialize the polynomial regression model
model_poly = LinearRegression()

# Perform K-Fold cross-validation with the polynomial features
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Collect cross-validation scores
cv_scores_poly = cross_val_score(model_poly, X_poly, y, cv=kf, scoring='r2')
print(f'Polynomial Regression Cross-validation scores: {cv_scores_poly}')
print(f'Average Polynomial CV score: {np.mean(cv_scores_poly):.4f}, Standard deviation: {np.std(cv_scores_poly):.4f}')

# Fit the polynomial model on the entire dataset to derive the final coefficients
model_poly.fit(X_poly, y)

# Extract coefficients and intercept for polynomial regression
poly_coefficients = model_poly.coef_
poly_intercept = model_poly.intercept_

print(f'Coefficients: {poly_coefficients}, Intercept: {poly_intercept}')

Polynomial Regression Cross-validation scores: [1. 1. 1. 1. 1.]
Average Polynomial CV score: 1.0000, Standard deviation: 0.0000
Coefficients: [ 0.00000000e+00 -1.23112897e-07  1.20000000e+01  2.85714287e-02
  2.00000000e+00 -1.05687681e-11], Intercept: 1.6574806068092585e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)


$\text{Size} = 0.0000166 + (12.0 \cdot \text{Temperature °C}) + (0.02857 \cdot \text{Mols KCL}^2) + (2.0 \cdot \text{Temperature °C}^2)$


Report on the score and interpret:

The model has a perfect score of 1, and a perfect standard deviation of 0. The intercept is nearly 0, at 0.000166. The high score is possible because the data set is artificial, and the model may have found the equation the data set was made with. The intercept being nearly zero is also possible due to the data set being artificial. The intercept being 0 also makes sense because a slime with 0 moles of KCl probably wouldn't be very large.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [38]:
# Load the cleaned CKD dataset
ckd_data = pd.read_csv('ckd_feature_subset.csv')  # Ensure the CSV file path is correct
print(ckd_data.head())

# Features (X) and Labels (y)
X_ckd = ckd_data.drop('Target_ckd', axis=1)  # Ensure this matches your target column name
y_ckd = ckd_data['Target_ckd']  # This is the target variable

# Split the dataset into training/testing (90% train, 10% test)
X_train_ckd, X_test_ckd, y_train_ckd, y_test_ckd = train_test_split(
    X_ckd, y_ckd, test_size=0.1, random_state=42
)

# Set up for 5-fold cross-validation
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Initialize model scores dictionary
model_scores = {
    'Model': [],
    'Average Accuracy': [],
    'Standard Deviation': []
}

# 1. Logistic Regression
log_reg = LogisticRegression(max_iter=1000)  # Increased iterations for convergence if needed
cv_scores_log_reg = cross_val_score(log_reg, X_ckd, y_ckd, cv=kf, scoring='accuracy')
model_scores['Model'].append('Logistic Regression')
model_scores['Average Accuracy'].append(np.mean(cv_scores_log_reg))
model_scores['Standard Deviation'].append(np.std(cv_scores_log_reg))
print(f'Logistic Regression - CV scores: {cv_scores_log_reg}')

# 2. Support Vector Machines
svc_model = SVC()
cv_scores_svc = cross_val_score(svc_model, X_ckd, y_ckd, cv=kf, scoring='accuracy')
model_scores['Model'].append('Support Vector Machines')
model_scores['Average Accuracy'].append(np.mean(cv_scores_svc))
model_scores['Standard Deviation'].append(np.std(cv_scores_svc))
print(f'Support Vector Machines - CV scores: {cv_scores_svc}')

# 3. k-Nearest Neighbors
knn_model = KNeighborsClassifier()
cv_scores_knn = cross_val_score(knn_model, X_ckd, y_ckd, cv=kf, scoring='accuracy')
model_scores['Model'].append('k-Nearest Neighbors')
model_scores['Average Accuracy'].append(np.mean(cv_scores_knn))
model_scores['Standard Deviation'].append(np.std(cv_scores_knn))
print(f'k-Nearest Neighbors - CV scores: {cv_scores_knn}')

# Create a DataFrame for easier visualization of scores
scores_df = pd.DataFrame(model_scores)

# Display the scores
print("\nModel Performance Summary:")
print(scores_df)

        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1
Logistic Regression - CV scores: [0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]
Support Vector Machines - CV scores: [0.90322581 1.         0.87096774 0.9        0.96666667]
k-Nearest Neighbors - CV scores: [0.93548387 1.         0.87096774 0.86666667 0.96666667]

Model Performance Summary:
                     Model  Average Accuracy  Standard Deviation
0      Logistic Regression          0.856559            0.066269
1  Support Vector Machines          0.928172            0.047601
2      k-Nearest Neighbors         

## Results and Conclusion for Classification Experiments

Results:
| Model                          | Average Accuracy | Standard Deviation |
|--------------------------------|------------------|--------------------|
| Logistic Regression            | 0.85656          | 0.06627            |
| Support Vector Machines        | 0.92817          | 0.04760            |
| k-Nearest Neighbors            | 0.92796          | 0.05244            |

Conclusion:
Overall the model with the best accuracy and the lowest standard deviation seems to be the Support Vector Machines model. However, the K-Nearest Neighbors had similar accuarcy and standard deviation. The Logistic Regression model seems to be clearly the worst with a signifigantly lower accuracy, and the highest standard deviation, indicating that logistic regression is the least accurate model, and also the least generalized model.

In [26]:
# 4. Neural Networks
# with 3 different configurations

# Initialize model scores dictionary
model_scores2 = {
    'Model': [],
    'Average Accuracy': [],
    'Standard Deviation': []
}

configs = [
    {'hidden_layer_sizes': (10,), 'activation': 'relu'},
    {'hidden_layer_sizes': (10,), 'activation': 'tanh'},
    {'hidden_layer_sizes': (20,), 'activation': 'relu'},
    {'hidden_layer_sizes': (20,), 'activation': 'tanh'},
    {'hidden_layer_sizes': (10, 10), 'activation': 'relu'},
    {'hidden_layer_sizes': (10, 10), 'activation': 'tanh'},
    {'hidden_layer_sizes': (20, 20), 'activation': 'relu'},
    {'hidden_layer_sizes': (20, 20), 'activation': 'tanh'}
]
   
for config in configs:
    nn_model = MLPClassifier(hidden_layer_sizes=config['hidden_layer_sizes'], activation=config['activation'], max_iter=3000)
    cv_scores_nn = cross_val_score(nn_model, X_ckd, y_ckd, cv=kf, scoring='accuracy')
    
    model_scores2['Model'].append(f'Neural Network {config["hidden_layer_sizes"]} {config["activation"]}')
    model_scores2['Average Accuracy'].append(np.mean(cv_scores_nn))
    model_scores2['Standard Deviation'].append(np.std(cv_scores_nn))
    print(f'Neural Network Configuration {config}: CV scores: {cv_scores_nn}')

# Create a DataFrame for easier visualization of scores
scores_df2 = pd.DataFrame(model_scores2)

# Display the scores
print("\nModel Performance Summary:")
print(scores_df2)


Neural Network Configuration {'hidden_layer_sizes': (10,), 'activation': 'relu'}: CV scores: [0.93548387 0.96774194 0.83870968 0.93333333 0.96666667]
Neural Network Configuration {'hidden_layer_sizes': (10,), 'activation': 'tanh'}: CV scores: [0.93548387 0.96774194 0.83870968 0.9        0.96666667]
Neural Network Configuration {'hidden_layer_sizes': (20,), 'activation': 'relu'}: CV scores: [0.93548387 0.96774194 0.83870968 0.93333333 0.96666667]
Neural Network Configuration {'hidden_layer_sizes': (20,), 'activation': 'tanh'}: CV scores: [0.90322581 0.96774194 0.83870968 0.93333333 0.96666667]
Neural Network Configuration {'hidden_layer_sizes': (10, 10), 'activation': 'relu'}: CV scores: [0.90322581 1.         0.90322581 0.96666667 0.96666667]
Neural Network Configuration {'hidden_layer_sizes': (10, 10), 'activation': 'tanh'}: CV scores: [0.96774194 0.96774194 0.87096774 0.96666667 0.96666667]
Neural Network Configuration {'hidden_layer_sizes': (20, 20), 'activation': 'relu'}: CV scores

## Results and Conclusion for Neural Network Experiments

Results:
| Model                            | Average Accuracy | Standard Deviation |
|----------------------------------|------------------|--------------------|
| Neural Network (10,) - ReLU      | 0.92839          | 0.04718            |
| Neural Network (10,) - Tanh      | 0.92172          | 0.04835            |
| Neural Network (20,) - ReLU      | 0.92839          | 0.04718            |
| Neural Network (20,) - Tanh      | 0.92194          | 0.04797            |
| Neural Network (10, 10) - ReLU   | 0.94796          | 0.03850            |
| Neural Network (10, 10) - Tanh   | 0.94796          | 0.03850            |
| Neural Network (20, 20) - ReLU   | 0.94172          | 0.04729            |
| Neural Network (20, 20) - Tanh   | 0.93505          | 0.04047            |



Conclusion:
The neural network models all showed high accuracies and low standard deviatons compared to the other models. The models with the highest accuarcy were the 10,10 ReLU and the 10,10 Tanh models. These models also had the lowest standard deviation. Since these models beat the models with 2 hidden layers of 20 nodes in accuracy and standard deviation, it indicates that models with more nodes per hidden layer aren't more accurate. Since the models with more hidden layers were always more accurate than the models with 1 hidden layer, it indiates that models with more hidden layers tend to be more accurate. Overall the models that seem to have the best accuaracy and lowerst standard deviation seem to be the ones with more hidden layers, and with lower amounts of nodes per hidden layer. The activation function does not seem to have a large effect on accuarcy or standard deviation.